teps/0007-conditions-beta.md

```diff
@@ -0,0 +1,609 @@
+
+Here's how a user can specify a `Guard` that executes on that `Result`: 
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: Task
+metadata:
+  name: file-exists
+spec:
+  params:
+    - name: path
+  workspaces:
+    - name: source
+  results:
+    - name: exists
+      description: boolean indicating whether the file exists
+  steps:
+  - name: check-file-exists
+    image: alpine
+    script: |
+      if [ -f $(workspaces.source.path)/$(params.path) ]; then
+        echo true | tee /tekton/results/exists
+      else
+        echo false | tee /tekton/results/exists
+      fi
+---
+api: tekton.dev/v1alpha1
+kind: Pipeline
+metadata: 
+    name: generate-file
+spec:
+  workspaces:
+    - name: source-repo
+  params:
+    - name: path
+      default: 'README.md'
+  tasks: 
+    - name: create-file
+      taskRef: 
+        name: create-readme-file
+      workspaces:
+        - name: source
+          workspace: source-repo
+    - name: file-exists
+      taskRef: 
+        name: file-exists
+      workspaces:
+        - name: source
+          workspace: source-repo
+    - name: echo-file-exists 
+      when:
+        - name: check-file-exists
+          input: '$(tasks.file-exists.results.exists)'
+          operator: In
+          values: ['true']
+      taskRef:
+        name: echo-file-exists
+```
+
+Other examples of `Guards` using `Match Expressions` are:
+
+```yaml
+name: branch-is-main
+input: $(params.branch)
+operator: In
+values: [‘main’]
+---
+name: branch-not-main
+input: $(params.branch)
+operator: NotIn
+values: [‘main’]
+---
+name: always-false
+input: ‘false’
+operator: In
+values: [‘’]
+```
+
+We can explore adding more `Operators` later if needed, such as `IsTrue`, `IsFalse`, `IsEmpty` and `IsNotEmpty`.
+
+### Skipping
+
+Provide more flexibility when a `Guard` executes as `False` by adding a field, `continueAfterSkip`, used to specify whether execute the `Tasks` that are ordering-dependent on the skipped guarded `Task`. The `continueAfterSkip` field defaults to `false`/`no` and users can set it to `true`/`yes` (case insensitive) to allow for execution of the rest of the branch. 
+
+A `Task` branch is made up of dependent `Tasks`, where there are two types of dependencies:
+- _Resource dependency_: based on resources needed from parent `Task`, which includes Workspaces, `Results` and Resources. 
+- _Ordering dependency_: based on runAfter which provides sequencing of `Tasks` when there may not be resource dependencies. 
+
```

Comment:
So far we have been using `Results` in `Params` and if specified, like you mentioned, result producing task becomes parent and executed before the task consuming it. In this case, a task is implicitly made dependent on the tasks specified in `Guard` (love the terminology) through `input`. Similarly if we have N separate guards, the guarded task is marked as dependent on all guards without specifying any `runAfter`.

Reply:
Using task results, this proposal is following the same `resource` dependency, but might be worth mentioning somewhere in our documentation, how Guard `input` reading task results introduces such implicit dependency.

Reply:
Thanks for elaborating about this @pritidesai ! Do you prefer _implicit_ or _explicit_ task dependencies?

I don't know if this is by design or not, @jerop @bobcatfish ?

My own opinion, is that I don't really know... no strong opinion...
...but I tend to prefer _explicit task ordering_ as with `runAfter`. But I don't think this is a big problem.

The "worst case" - were this dependency is a bit unclear to the user might be a pipeline like this:

```
         (b) — (e) — (f) — (g) — (h) — (i)
       / 
   (a) 
       \
         (c) — (d) — (s) — (t)
```
e.g. task (a) do `git-clone`, and the upper branch builds backend (including integration testing), the lower branch builds frontend.

A condition is set on `(s)` - if `(i)` has huge artifact size and - if `(d)` has huge artifact size - print message on Slack.

This is effectively this dependency graph - but it is **not very clear to the user** - (unless `runAfter:` is required and validated)  - but it is more heavy weight for the user to have to declare this as well.

```
         (b) — (e) — (f) — (g) — (h) — (i)
       /                                 \
   (a)                                    \
       \                                   \
         (c) — (d) —————————————————————— (s) — (t)
```

I don't know what is best - but as you say - it is good to be clear about what the intention is.

Reply:
This is great example @jlpettersson, yes explicit task ordering is specified with `runAfter` like @jerop classified it as ordering dependency. Task results (causes implicit dependency) classified as resource dependency. Yup, it will be helpful to document 🙏

Reply:
> es explicit task ordering is specified with runAfter like @jerop classified it as ordering dependency. Task results (causes implicit dependency) classified as resource dependency.

@jlpettersson it is somewhat by design - i say "somewhat" because it didnt become clear to us until maybe the last 6 months but basically it seems useful to be able to distinguish between these two types of dependencies, and it's not so much that they are explicit vs implicit but that they are, like @pritidesai  said:

1. Ordering dependency (runAfter)
2. Resource dependency (result, pipeline resource, one day workspaces)

Having the controller be able to be aware of the difference between these 2 cases is useful - e.g. specifically in this case we know that (2) + continueAfterSkip is bad; if we only had (1) we wouldn't know that

Back to implicit vs explicit tho: we had some debate about the [`from` syntax that PipelineResources use](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#using-the-from-parameter) vs. the variable replacement which implies ordering that we used for results (e.g. `$(tasks.foo.results.bar)`)

Personally I prefer the explicit `from` - when you see `from`, you know that ordering is being required as well as a value being used. [this is the doc where we discussed it](https://docs.google.com/document/d/12fOveI3_bVS1on6yCZK3rY5vki5EtvYFzNr350UhvwI/edit#heading=h.gb8t5489ktdt) tho im not sure if this makes it clear why we went with variable interpolation instead of from. i think adding support for "from" could still be on the table but id be surprised if we removed the variable interpolation option